### ISSUE TITLE
Release types

### ISSUE THREAD

**Issue by OxygenCobalt** [OxygenCobalt/Auxio#158]
This issue is about implementing release type functionality. Release Type is an niche tag that signifies whether a grouping of songs is an Album, EP, Single, or other. Since Auxio will soon be able to extract arbitrary tags with #128, this seems like a good feature to add.

This is how the feature will likely work:
- Auxio will only accept the **primary** tags outlined by MusicBrainz. These are `album`, `ep`, and `single`. Secondary tags or arbitrary tags will currently be ignored.
- In the artist view, albums will be separated by release type.
- In the album view, albums will have their release type next to their song count, so "EP / 10 Songs"
- If #128 is not enabled, Auxio's current behavior will stay as is, as I cannot backport this behavior to the android media database.

**Comment by rhynec** [comment]
@OxygenCobalt, do you think you could also support the "live" album tag? I saw you added support for some additional types, so I thought I might ask.

**Comment by OxygenCobalt** [comment]
Yeah, I added support for compilations and soundtracks, but not some of the other types like Live Recordings, Mixtapes, or Remixes. I can easily add any type (within reason), you let me know what types you want @rhynec. I'll go and implement the live tag.

**Comment by OxygenCobalt** [comment]
Yeah, after doing some thinking, this is the mapping I'm planning to use @rhynec . I've excluded some of the release types, but this is because Auxio is a music player whereas MusicBrainz is a general audio metadata standard.

```
Album -> album, invalid
EP -> EP, EP + invalid
Single -> single, single + invalid 
Compilation -> album + compilation, ep + compilation, single + compilation, compilation
Soundtrack -> album + soundtrack, ep + soundtrack, single + soundtrack, soundtrack

Remixes:
Remix Album -> album + remix, remix
Remix EP -> ep + remix
Remix Single -> ep + single

Live recordings:
Live Album -> album + live, live
Live EP -> ep + live
Live Single -> single + live

Mixtape -> album + mixtape, ep + mixtape, single + mixtape, mixtape
```

**Comment by rhynec** [comment]
Cool! Seems like a clean solution to handle different permutation of primary/secondary tags. Looking forwards to trying out the new feature.

**Comment by OxygenCobalt** [comment]
Just to let you know @rhynec, Release Type functionality is probably going to change in the next version.

I've recently figured a way to extract multiple values, at least from ID3v2.4 tags. Plus, I learned that MusicBrainz does not separate release types with +, but rather / or ;. 

In general, the next version will likely only expect multi-value tags normally. This produces sensible behavior without mangling tags. However, there will be a menu to enable particular separators from a list of `+ , ; / and &`. There will probably also be escaping functionality such as "\+" or "++" becoming "+" and not being parsed as several tags.

**Comment by rhynec** [comment]
Beat me to it, I was planning on suggesting this since you were working on multi artist/genre support. 

Out of curiosity, how do you plan to handle albums tagged with multiple secondary types, like compilation/live or compilation/remix? Would you just take the first secondary tag that Auxio finds, or would there be some priority system in the code (like in the previous examples, giving priority to live/remix over compilation)?

**Comment by OxygenCobalt** [comment]
The initial release types implementation was an MVP, and I decided not to add compilation subtypes as part of that. It should be an extremely easy implementation now however, all I need to add are the following rules:

```
album + compilation + live -> Live compilation
album + compilation + remix -> Remix compilation
album + compilations + dj-mix -> DJ Mix
album + compilation + soundtrack -> Soundtrack
```

@rhynec

**Comment by chocmake** [comment]
Is there any way to search/filter by release type? Eg: searching `ep` returns no results for an EP that has `EP` as its releasetype tag. It does however correctly group releases with them when an artist's name is clicked.

As my entire collection is tagged with them (custom values but thankfully most match those that have been included above) it'd be useful to be able to find results of albums and songs that match a particular type (eg: soundtrack, EP, single, etc).

**Comment by OxygenCobalt** [comment]
> any chance you could add support for demo?

Sorry for responding so late to this @xfzv. IIRC this isn't music-related, so I didn't include such in Auxio's subset. Can you explain it?

> Is there any way to search/filter by release type? Eg: searching ep returns no results for an EP that has EP as its releasetype tag. It does however correctly group releases with them when an artist's name is clicked.

Not really @chocmake. Release types are largely "sugar" in Auxio, not categorizations. This is a bit of a spotify-ism, yes, but there are so many different release types that treating them all as categories would result in a highly cluttered UI. Overloading search like that also isn't desirable, as it would probably make search results unusable for certain plaintext queries.

**Comment by OxygenCobalt** [comment]
Got it. I'll add it then @xfzv.

**Comment by chocmake** [comment]
> but there are so many different release types to the point where treating them all as categories would result in a highly cluttered UI. Overloading search like that also isn't desirable, as it would probably make search results unusable for certain plaintext queries.

For example there's no other way to show all soundtracks, if one wanted to filter to only that type of release. To have any matches it would require `soundtrack` to appear in the title of a release but that isn't always the case for album titles that use `OST` or `Music of` instead. One could use abuse genre tags instead but then that's kind of redundant given releasetype exists.

If such matching weren't triggered by a search query at least being able to toggle a filter that limited items to a certain release type on demand would be useful, just putting it out there :)

**Comment by OxygenCobalt** [comment]
Sorry for taking awhile to respond @chocmake. I think a chip pattern would be much more suitable for being able to control release types (Tracked in #329), but again, I would need to find a good way to avoid making it cluttered. 

It could also work in the library if I were to make the "Albums" tab into a "Releases" tab that then allows individual filtering by release types using the chip pattern. The issue with that is how that would be a lot more alienating to users with simple music libraries, who expect everything to be Albums. Perhaps I add a "nerd mode" to settings that allows you to flip between the two views, lol.

**Comment by chocmake** [comment]
> Sorry for taking awhile to respond @chocmake. I think a chip pattern would be much more suitable for being able to control release types (Tracked in #329)

No prob; sounds good.

> It could also work in the library if I were to make the "Albums" tab into a "Releases" tab that then allows individual filtering by release types using the chip pattern. The issue with that is how that would be a lot more alienating to users with simple music libraries [...] perhaps it could be dynamically enabled if a library is detected to have non-albums.

Yeah, if the library were scanned and non-albums detected then dynamically making the filtering options visible makes sense since it would avoid unnecessarily filtering for non-present release types.

From a UI PoV with the search I can see the challenge of how to place more options, since I'm guessing the reason it's a drop-down currently is since the keyboard is visible when the search field has been typed in, so the larger bottom-up Sort By tray used for the regular tabs isn't used here.

An alternative choice I've seen RedReader adopt for a similar scenario is a multi-level drop-down filter. Where upon clicking a sort filter option in the list, such as sort by *Top*, it replaces the list with different options (hourly, today, weekly, etc). Its implementation lacks any transition animation (like a slide) between such expanding options so it doesn't feel slick (not sure if it's any inherent limitation or they just didn't go the extra mile).

Otherwise perhaps if it were a modal popup instead maybe some different method could be used. Does Android have a way of temporarily hiding the keyboard when a modal is triggered? As that could afford back some extra vertical height which is the main drawback of figuring out how to place the options otherwise.

**Comment by OxygenCobalt** [comment]
> From a UI PoV with the search I can see the challenge of how to place more options, since I'm guessing the reason it's a drop-down currently is since the keyboard is visible when the search field has been typed in, so the larger bottom-up Sort By tray used for the regular tabs isn't used here.

Not really. That's just the way I designed it originally. I honestly don't want to use a modal for something like this, you will want to re-sort much less often compared to re-filtering something. 

I'm assuming perhaps the chip pattern would work something like:
- In hypothetical "Releases" tab: Albums, EPs, Singles, Compilations, Live, Remixes, so on
- In search: Songs, Releases, Artists, Genres, Playlists. The Releases button then opens a little dropdown (not a dialog) that then allows you to select by Albums, EPs, Singles, Compilations, etc.

These both sound like good ideas, assuming I can dynamically enable them.

**Comment by chocmake** [comment]
> * In search: Songs, Releases, Artists, Genres, Playlists. The Releases button then opens a little dropdown (not a dialog) that then allows you to select by Albums, EPs, Singles, Compilations, etc.

I edited it out of my prior post for brevity but RedReader does something similar for its subreddit sort drop-down. In that some items in the list are a standard state while pressing some like (sort by) *Top* replaces the list with further options (selecting one then closes the filter drop-down). In their case they didn't implement any transition animation between the views so it's not particularly slick-feeling but it works. Not sure if that's what you had in mind.

**Comment by OxygenCobalt** [comment]
> I edited it out of my prior post for brevity but RedReader does something similar for its subreddit sort drop-down. In that some items in the list are a standard toggle state while pressing some like (sort by) Top replace the list with further options (selecting one then closes the filter drop-down, as they're radio options). In their case they didn't implement any transition animation between the views so it's not particularly slick-feeling but it works. Not sure if that's what you had in mind.

I think I get what you mean, but a screenshot would be useful here.

**Comment by chocmake** [comment]
Here's a video capture. (I'm reminded there is a transition animation but the sudden offset gave me a different impression)


https://github.com/OxygenCobalt/Auxio/assets/34178938/6d1edc37-13e2-411d-aa7b-3098b389dca7

**Comment by OxygenCobalt** [comment]
Ah. Yes, it's very much like that, but combined with the chip idiom, like below:


https://github.com/OxygenCobalt/Auxio/assets/65370175/262915a4-cdad-4b98-a563-7a7e0619597c

**Comment by chocmake** [comment]
Ah, nice. That example with horizontally scrollable chips would also solve the limited space issue as well if there was any overflow.

**Comment by OxygenCobalt** [comment]
It would also allow multi-selecting filters. For example, you could arbitrarily enable "Artists, Songs, EPs, Singles" and filter out everything else.

Note that #329 will probably not have this, instead just having an "Albums" chip (Again due to the terminology issues). I'm going to spin this off into a new issue.

**Comment by OxygenCobalt** [comment]
Spun off into #593.
